File: layers/Sigmoid.m
classdef Sigmoid < BaseLayer
    %Sigmoid Summary of this class goes here
    % Reference: https://github.com/leonardoaraujosantos/DLMatFramework/blob/master/learn/cs231n/assignment2/cs231n/layers.py
    
    properties (Access = 'protected')
        weights
        biases
        activations
        gradients
        config
        previousInput
        name
        index
        activationShape
        inputLayer
        weightShape
    end
    
    methods (Access = 'public')
        function [obj] = Sigmoid(name, index, inLayer)
            obj.name = name;
            obj.index = index;
            obj.inputLayer = inLayer;
            % Relu does not change the shape of it's output
            obj.activationShape = obj.inputLayer.getActivationShape();
            obj.weightShape = [];
        end
        
        function [activations] = ForwardPropagation(obj, input, weights, bias)
            tic;
            
            obj.previousInput = input;
            activations = (1./(1+exp(-input)));
            obj.activations = activations;
            
            % Get execution time
            obj.executionTime = toc;
        end
        
        function [gradient] = BackwardPropagation(obj, dout)
            dout = dout.input;
            t = (1./(1+exp(-obj.previousInput)));
            d_sigm  = (t .* (1 - t));
            dx = dout .* d_sigm;
            gradient.input = dx;
            
            % Cache gradients
            obj.gradients = gradient;
            
            if obj.doGradientCheck
                evalGrad = obj.EvalBackpropNumerically(dout);
                diff_Input = sum(abs(evalGrad.input(:) - gradient.input(:)));                
                diff_vec = [diff_Input]; 
                diff = sum(diff_vec);
                if diff > 0.0001
                    msgError = sprintf('%s gradient failed!\n',obj.name);
                    error(msgError);
                else
                    %fprintf('%s gradient passed!\n',obj.name);
                end
            end
        end
        
        function gradient = EvalBackpropNumerically(obj, dout)
            % Fully connected layers has 3 inputs so we have 3 gradients
            sigm_x = @(x) obj.ForwardPropagation(x,obj.weights, obj.biases);            
            
            % Evaluate
            gradient.input = GradientCheck.Eval(sigm_x,obj.previousInput, dout);            
        end
    end
    
end

File: utils/optimization_functions/cppPrj/layers/sigmoid.cpp
#include "sigmoid.h"

Sigmoid::Sigmoid(const string &name, shared_ptr<BaseLayer> inLayer){
    m_inputLayer = inLayer;
    m_name = name;

    // Sigmoid does not change the shape of it's input
    if (m_inputLayer != nullptr){
        m_activationShape = m_inputLayer->GetActivationShape();
    }
}

Tensor<float> Sigmoid::ForwardPropagation(const Tensor<float> &input){
    Tensor<float> activation = 1.0/(1.0+MathHelper<float>::Exp(-input));

    // Cache information for backpropagation
    m_previousInput = input;
    m_activation = activation;

    return activation;
}

LayerGradient<float> Sigmoid::BackwardPropagation(const LayerGradient<float> &dout){
    Tensor<float> t = 1.0/(1.0+MathHelper<float>::Exp(-m_previousInput));
    Tensor<float> d_sigm = t.EltWiseMult(1.0-t);
    Tensor<float> dx = dout.dx.EltWiseMult(d_sigm);

    // Create empty tensors
    Tensor<float> dWeights;
    Tensor<float> dBias;
    LayerGradient<float> gradient{dx,dWeights,dBias} ;

    // Cache gradients
    m_gradients = gradient;

    return gradient;
}
